### PR TITLE
Add RC option for quadplane manual forward throttle

### PR DESCRIPTION
**Pull request by kd0aij**
To test, set RCx_OPTION to 209 where x is the RC input channel controlling forward throttle
and set parameter Q_FWD_THR_MAX to the maximum value in percent. For the Convergence tilt vectored tricopter frame, 25% is a reasonable maximum.

### CODE REVIEW

File: libraries/RC_Channel/RC_Channel.h [hunk 1]
```diff
@@ -181,6 +181,7 @@ class RC_Channel {
         DISARM =              81, // disarm vehicle
         Q_ASSIST =            82, // disable, enable and force Q assist
         ZIGZAG_Auto =         83, // zigzag auto switch
+        FWD_THR =             84, // VTOL manual forward throttle
```
Comment (IamPete1) [inline libraries/RC_Channel/RC_Channel.h:184]: I have been adding manual control analog inputs at the end to line up with where the four main ones will end up when we finally kill RCMAP. So this would be 209

File: ArduPlane/quadplane.h [hunk 2]
```diff
@@ -335,6 +335,11 @@ class QuadPlane
     // manual throttle curve expo strength
     AP_Float throttle_expo;
 
+    // manual forward throttle input
+    AP_Int8  fwd_thr_chan;
```
Comment (IamPete1) [inline ArduPlane/quadplane.h:339]: I think this is not needed anymore

File: ArduPlane/tiltrotor.cpp [hunk 3]
```diff
@@ -92,29 +92,48 @@ void QuadPlane::tiltrotor_continuous_update(void)
...
         tiltrotor_slew(0);
         return;
     }
 
+    // if not in assisted flight and in QACRO, QSTABILIZE or QHOVER mode
+    if (!assisted_flight &&
+        (plane.control_mode == &plane.mode_qacro ||
+         plane.control_mode == &plane.mode_qstabilize ||
+         plane.control_mode == &plane.mode_qhover)) {
+        if (rc_fwd_thr_ch == nullptr) {
```
Comment (IamPete1) [inline ArduPlane/tiltrotor.cpp:126]: I think is better to just re-get the pointer each time, then it works straight away without a reboot

File: ArduPlane/quadplane.cpp [hunk 4]
```diff
@@ -2910,24 +2920,50 @@ void QuadPlane::Log_Write_QControl_Tuning()
   reduces the need for down pitch which reduces load on the vertical
   lift motors.
  */
-int8_t QuadPlane::forward_throttle_pct(void)
-{
+int8_t QuadPlane::forward_throttle_pct(bool tiltrotor)
+{
+    // if not a tiltrotor, disable forward motor:
+    if (!tiltrotor) {
```
Comment (IamPete1) [inline ArduPlane/quadplane.cpp:2926]: I think we should be able to make this work for all quad planes

File: libraries/RC_Channel/RC_Channel.h (hunk 1 above)
Comment (kd0aij) [reply libraries/RC_Channel/RC_Channel.h:184]: thanks, changed to 209

File: ArduPlane/quadplane.h (hunk 2 above)
Comment (kd0aij) [reply ArduPlane/quadplane.h:339]: good catch, removed

File: ArduPlane/tiltrotor.cpp (hunk 3 above)
Comment (kd0aij) [reply ArduPlane/tiltrotor.cpp:126]: I moved the find call from init to the one_second_loop, so we're not doing a linear search in the fast loop, but can still change it without rebooting.

File: ArduPlane/quadplane.cpp (hunk 4 above)
Comment (kd0aij) [reply ArduPlane/quadplane.cpp:2923]: That comment may be confusing... manual forward throttle does work for standard quadplanes (SITL tested on the RF8 BigStik quadplane), just not when disarmed.

The change in behavior here is only that tiltrotor motors can tilt forward when disarmed.  This is a good thing because you don't want them to suddenly tilt forward on arming since some planes (like the Convergence) have no clearance to the ground.  
This method is only called with tiltrotor==true from tiltrotor_continuous_update, and that method only affects tilt servos, not throttles.

For a non-tiltrotor, we must return zero here when disarmed or spooled down, otherwise props could spin when they shouldn't.

**Comment by Hwurzburg** [comment]
just a nit...but wouldn't the name Q_FWD_MANTHR_MAX (still fits I think) or Q_FWD_RC_THR_MAX be more descriptive and not lead one to think this might limit throttle when using Q_VFWD_GAIN?

File: ArduPlane/quadplane.h [hunk 5]
```diff
@@ -101,7 +101,7 @@ class QuadPlane
     }
 
     // return desired forward throttle percentage
-    int8_t forward_throttle_pct(void);        
+    int8_t forward_throttle_pct(bool tiltrotor);
```
Comment (IamPete1) [inline ArduPlane/quadplane.h:104]: you could default this to false and maybe make the bool return_disarmed or something. tilltrotor only makes sense from the context of the function calls but not within the function itself

Comment (kd0aij) [reply ArduPlane/quadplane.h]: saved a few lines of code and (I think) eliminated confusion by moving the arming check up to the caller (this is only called from 2 places)

**Comment by kd0aij** [comment]
@Hwurzburg Q_FWD_MANTHR_MAX is 16 characters, which is exactly the max allowed length 👍

File: libraries/RC_Channel/RC_Channel.cpp [hunk 6]
```diff
@@ -92,7 +92,7 @@ const AP_Param::GroupInfo RC_Channel::var_info[] = {
...
+    // @Values{Copter}: 0:Do Nothing, 2:Flip, 3:Simple Mode, 4:RTL, 5:Save Trim, 7:Save WP, 9:Camera Trigger, 10:RangeFinder, 11:Fence, 13:Super Simple Mode, 14:Acro Trainer, 15:Sprayer, 16:Auto, 17:AutoTune, 18:Land, 19:Gripper, 21:Parachute Enable, 22:Parachute Release, 23:Parachute 3pos, 24:Auto Mission Reset, 25:AttCon Feed Forward, 26:AttCon Accel Limits, 27:Retract Mount, 28:Relay On/Off, 29:Landing Gear, 30:Lost Copter Sound, 31:Motor Emergency Stop, 32:Motor Interlock, 33:Brake, 34:Relay2 On/Off, 35:Relay3 On/Off, 36:Relay4 On/Off, 37:Throw, 38:ADSB Avoidance En, 39:PrecLoiter, 40:Proximity Avoidance, 41:ArmDisarm, 42:SmartRTL, 43:InvertedFlight, 46:RC Override Enable, 47:User Function 1, 48:User Function 2, 49:User Function 3, 52:Acro, 55:Guided, 56:Loiter, 57:Follow, 58:Clear Waypoints, 60:ZigZag, 61:ZigZag SaveWP, 62:Compass Learn, 65:GPS Disable, 66:Relay5, 67:Relay6, 68:Stabilize, 69:PosHold, 70:AltHold, 71:FlowHold, 72:Circle, 73:Drift, 76:Standby Mode, 78:RunCam Control, 79:RunCam OSD Control, 80:Viso Align, 81:Disarm, 83:ZigZag Auto, 84: Forward Throttle, 100:KillIMU1, 101:KillIMU2, 102:Camera Mode Toggle
```
Comment (IamPete1) [inline libraries/RC_Channel/RC_Channel.cpp:95]: I think you meant to add to plane rather than copter

Comment (kd0aij) [reply libraries/RC_Channel/RC_Channel.cpp:95]: I"m going to need a bigger font :)
Also changed to 209 and fixed the order of the Plane options

File: ArduPlane/quadplane.cpp [hunk 7]
```diff
@@ -2910,24 +2917,41 @@ void QuadPlane::Log_Write_QControl_Tuning()
...
+      Unless an RC channel is assigned for manual forward throttle control,
+      we don't use forward throttle in QHOVER or QSTABILIZE as they are the primary
       recovery modes for a quadplane and need to be as simple as
-      possible. They will drift with the wind
+      possible. They will drift with the wind.
     */
-    if (!in_vtol_mode() ||
-        !motors->armed() ||
-        vel_forward.gain <= 0 ||
+    if (plane.control_mode == &plane.mode_qacro ||
```
Comment (tridge) [inline ArduPlane/quadplane.cpp:2928]: this change removes some of the protections we had, specifically the ground idle check, the motors armed check and the vtol mode check

File: ArduPlane/quadplane.cpp [hunk 8]
```diff
@@ -2910,24 +2917,41 @@ void QuadPlane::Log_Write_QControl_Tuning()
...
+            float fwd_thr = (1.0f + rc_fwd_thr_ch->norm_input()) / 2;
+
+            // set forward throttle to fwd_thr_max * (manual input + mix): range [0,100]
+            fwd_thr = constrain_float(fwd_thr_max, 0, 100) * constrain_float(fwd_thr, 0, 1);
+            return fwd_thr;
+        }
+    }
+
+    /*
+      in qautotune mode or modes without a velocity controller
```
Comment (tridge) [inline ArduPlane/quadplane.cpp:2945]: comment wrong

File: ArduPlane/servos.cpp [hunk 9]
```diff
@@ -504,9 +504,14 @@ void Plane::set_servos_controlled(void)
         // manual pass through of throttle while in GUIDED
         SRV_Channels::set_output_scaled(SRV_Channel::k_throttle, get_throttle_input(true));
     } else if (quadplane.in_vtol_mode()) {
-        // ask quadplane code for forward throttle
-        SRV_Channels::set_output_scaled(SRV_Channel::k_throttle, 
-            constrain_int16(quadplane.forward_throttle_pct(), min_throttle, max_throttle));
+        int16_t fwd_thr = 0;
+        // if armed and not spooled down ask quadplane code for forward throttle
+        if (quadplane.motors->armed() &&
```
Comment (tridge) [inline ArduPlane/servos.cpp:509]: why is this moved here?

File: ArduPlane/quadplane.cpp [hunk 10]
```diff
@@ -484,6 +484,13 @@ const AP_Param::GroupInfo QuadPlane::var_info2[] = {
     // @User: Standard
     AP_GROUPINFO("TAILSIT_GSCMIN", 18, QuadPlane, tailsitter.gain_scaling_min, 0.4),
 
+    // @Param: FWD_MANTHR_MAX
```
Comment (tridge) [inline ArduPlane/quadplane.cpp:487]: the param name and description don't make it clear that this is related to tilting rotors

File: ArduPlane/quadplane.cpp (hunk 7 above)
Comment (kd0aij) [reply ArduPlane/quadplane.cpp:2928]: As the commit message says, the intent is to move the checks to the caller
unless I screwed up the logic...

File: ArduPlane/quadplane.cpp (hunk 8 above)
Comment (kd0aij) [reply ArduPlane/quadplane.cpp:2945]: looks right to me; what is the issue?

File: ArduPlane/servos.cpp (hunk 9 above)
Comment (kd0aij) [reply ArduPlane/servos.cpp:509]: because this is the only place where the check is needed

File: ArduPlane/quadplane.cpp (hunk 10 above)
Comment (kd0aij) [reply ArduPlane/quadplane.cpp:487]: it applies to all quadplanes

Comment (tridge) [reply ArduPlane/quadplane.cpp:487]: ok. i think the description should reference the RCn_OPTION being set to create a fwd throttle option

Comment (kd0aij) [reply ArduPlane/quadplane.cpp:487]: expand param desc and add arming check

**Comment by tridge** [comment]
needs an arming check for zero throttle

**Comment by kd0aij** [comment]
@tridge @WickedShell  I discovered when adding the pre-arm check that throttle is ignored by default (RC_OPTIONS defaults to zero which means arming_check_throttle returns false)
Is this the intended behavior?  seems backwards

**Comment by WickedShell** [comment]
@kd0aij yeah... I was unable to carry the opinion we should check it by default on the dev call, so it was an opt in arming check, while roll/pitch/yaw are opt out....

**Comment by kd0aij** [comment]
@tridge squashed and retested in SITL
new throttle pre-arm check works if enabled in RC_OPTIONS (non-default)
Ready to merge unless you'd prefer the arming check active by default.

**Comment by kd0aij** [comment]
@peterbarker Which "comments from Pete are unresolved"?
I thought I had addressed all of those from @IamPete1

**Comment by peterbarker** [comment]
@kd0aij I usually hit "Resolve Conversation" in the github interface if the fix has been done.

File: ArduPlane/quadplane.cpp [hunk 11]
```diff
@@ -2910,24 +2917,41 @@ void QuadPlane::Log_Write_QControl_Tuning()
...
+            float fwd_thr = rc_fwd_thr_ch->percent_input();
+
+            // set forward throttle to fwd_thr_max * (manual input + mix): range [0,100]
+            fwd_thr *= .01f * constrain_float(fwd_thr_max, 0, 100);
+            return fwd_thr;
+        }
+    }
+
+    /*
+      in qautotune mode or modes without a velocity controller
```
Comment (peterbarker) [inline ArduPlane/quadplane.cpp:2945]:
```suggestion
      Vehicle is in in a mode without a velocity controller
```

... on the assumption that qautotune *does* have a velocity controller.

Comment (kd0aij) [reply ArduPlane/quadplane.cpp:2945]: this zeros out the forward throttle if in qautotune or if the relevant gain is zero
We want zero airspeed in qautotune, so the forward motor stays shut down.

**Comment by kd0aij** [comment]
@peterbarker wow, there were a lot of those buttons :)
Thanks for looking at this and #14003...

**Comment by Hwurzburg** [comment]
the throttle arming check should pass if is_flying is true.....

**Comment by kd0aij** [comment]
After discussion with @Hwurzburg and @tridge on pre-arm throttle checks, I have changed the check to a threshold of 5% 
and changed it to be on by default: the option is now to skip the check instead.  The check is also skipped if AP_Notify::flags.flying is true to allow for in-air rearming.
*** Note that the RPY controls neutral check is not skipped when flying; perhaps it should also be skipped when already flying?

**Comment by kd0aij** [comment]
@andyp1per This is repeatedly failing step 4 of the fly_glyro_fft test, but I can't see why this PR would break it...

**Comment by andyp1per** [comment]
@kd0aij - it failed to takeoff, that's why the test failed - which sounds related?

**Comment by andyp1per** [comment]
The "wanted 4" thing is because it failed to become armable

**Comment by kd0aij** [comment]
@andyp1per Thanks... how did you determine that? Do you have to run the test locally to see the buildlogs?

**Comment by andyp1per** [comment]
I'm expert at knowing what the failures look like ;)
